protocols/gossipsub/src/error.rs

```diff
@@ -87,20 +88,20 @@ impl From<SigningError> for PublishError {
 }
 
 /// Errors that can occur in the protocols handler.
-#[derive(Debug)]
+#[derive(Debug, Error)]
 pub enum GossipsubHandlerError {
-    /// The maximum number of inbound substreams created has been exceeded.
+    #[error("The maximum number of inbound substreams created has been exceeded.")]
     MaxInboundSubstreams,
-    /// The maximum number of outbound substreams created has been exceeded.
+    #[error("The maximum number of outbound substreams created has been exceeded.")]
     MaxOutboundSubstreams,
-    /// The message exceeds the maximum transmission size.
+    #[error("The message exceeds the maximum transmission size.")]
     MaxTransmissionSize,
-    /// Protocol negotiation timeout.
+    #[error("Protocol negotiation timeout.")]
     NegotiationTimeout,
-    /// Protocol negotiation failed.
+    #[error("Protocol negotiation failed.")]
     NegotiationProtocolError(ProtocolError),
-    /// IO error.
-    Io(std::io::Error),
+    #[error("Failed to encode or decode")]
+    Codec(Error),
```

Comment:
```suggestion
    Codec(#[source] Error),
```

This should fix your compile error with `?`!